web/accounts/oauth.go

```diff
@@ -165,14 +165,43 @@ func redirect(c echo.Context) error {
 		}
 	}
 
-	if err := couchdb.CreateDoc(i, acc); err != nil {
-		return err
+	if connID := c.QueryParam("connection_id"); connID != "" {
+		if existingAccount, err := findAccountWithSameConnectionID(i, connID); err == nil {
+			acc = existingAccount
+		}
+	}
+
+	if acc.ID() == "" {
+		if err := couchdb.CreateDoc(i, acc); err != nil {
+			return err
+		}
 	}
 
 	c.Set("instance", i.WithContextualDomain(c.Request().Host))
 	return redirectToApp(c, i, acc, clientState, slug, "")
 }
 
+func findAccountWithSameConnectionID(inst *instance.Instance, connectionID string) (*account.Account, error) {
+	var accounts []*account.Account
+	req := &couchdb.AllDocsRequest{Limit: 1000}
+	err := couchdb.GetAllDocs(inst, consts.Accounts, req, &accounts)
```

Comment:
When I first look to implement this feature, I saw that our couchdb package was not supporting the $in operator (https://github.com/cozy/cozy-stack/blob/master/pkg/couchdb/mango/query.go) I know that this $in operator is pretty slow, but I never expected that `all_docs` then a mapping was quicker than a mango query with a $in.

Reply:
If we had millions+ of accounts, mango would be faster, but with a very small number of accounts, all_docs wins by a lot.